File: src/main/java/j/template_method/inheritance/StringDisplay.java
package j.template_method.inheritance;

public final class StringDisplay extends AbstractDisplay {
  private final String string;
  private final int width;

  // コンストラクタ
  public StringDisplay(String string) {
    this.string = string;
    this.width = string.length();
  }

  @Override
  public void open() {
    printLine();
  }

  @Override
  public void print() {
    System.out.println("|" + string + "|");
  }

  @Override
  public void close() {
    printLine();
  }

  private void printLine() {
    System.out.print("+");
    for (int i = 0; i < width; i++) {
      System.out.print("-");
    }
    System.out.println("+");
  }
}

File: src/main/kotlin/k/template_method/delegate/StringDisplay.kt
package k.template_method.delegate

import j.template_method.delegate.DisplayDelegate

class StringDisplay(private val string: String) : DisplayDelegate {
  private val width: Int = string.length

  override fun open() {
    printLine()
  }

  override fun print() {
    println("|$string|")
  }

  override fun close() {
    printLine()
  }

  private fun printLine() {
    print("+")
    for (i in 0 until width) {
      print("-")
    }
    println("+")
  }
}
